### PR TITLE
Do the actual swap

### PR DESCRIPTION
**Pull request by fiksn**

### CODE REVIEW

File: plugins/boltz/db.go [hunk 1]
```diff
@@ -21,11 +23,17 @@ type BoltzDB struct {
 }
 
 func (b *BoltzDB) Connect(dbPath string) error {
+	if _, err := os.Stat(dbPath); errors.Is(err, os.ErrNotExist) {
```
Comment (dcrystalj) [inline plugins/boltz/db.go:26]: why is this needed. bolthold creates database if it does not exist.

File: plugins/boltz/swap.go [hunk 2]
```diff
@@ -4,10 +4,94 @@ import (
...
+		invoice.paid: if paying the invoice was successful
+		invoice.failedToPay: if paying the invoice failed. In which case the locked up onchain coins should be refunded
+		transaction.claimed
+	*/
+
+}
+
+// CalcFundsToReceive calculates how much funds you would receive
+func (b *Plugin) CalcFundsToReceive(ctx context.Context, reverse bool, sats uint64) (uint64, int64, error) {
+	if b.LnAPI == nil {
```
Comment (dcrystalj) [inline plugins/boltz/swap.go:30]: Can we remove this part. It is not "CalcFudnsFN" responsibility to initialize lnapi. We should expect it here to be set & ready. otherwise leave it to crash

File: plugins/boltz/swap.go [hunk 3]
```diff
@@ -4,10 +4,94 @@ import (
...
+		// reverse submarine requires huge outbound liquidity
+		liq, err := b.GetNodeLiquidity(ctx)
+		if err != nil {
+			return 0, 0, err
+		}
+
+		if liq.OutboundSats < sats {
+			return 0, int64(liq.OutboundSats - sats), fmt.Errorf("insufficent off-chain funds")
+		}
+	}
```
Comment (dcrystalj) [inline plugins/boltz/swap.go:59]: This is very complex function which should be split by "reverse" IMO.
```
FundsToReceiveSwap() {
  GetOnChainFunds()
  CheckFeeWithinLimits()
  fees = ...
  return amount, nil
}

FundsToReceiveReversedSwap() {
  GetNodeLiquidity()
  CheckFeeWithinLimits()
  fees = ..
  return amount, nil
}
```

File: plugins/boltz/swap.go [hunk 4]
```diff
@@ -4,10 +4,94 @@ import (
...
+		once the said transaction is included in a block (or found in the mempool in case of 0-confirmation) Boltz will try to pay the invoice provided by the user in order to claim the onchain coins
+		invoice.paid: if paying the invoice was successful
+		invoice.failedToPay: if paying the invoice failed. In which case the locked up onchain coins should be refunded
+		transaction.claimed
+	*/
+
+}
+
+// CalcFundsToReceive calculates how much funds you would receive
+func (b *Plugin) CalcFundsToReceive(ctx context.Context, reverse bool, sats uint64) (uint64, int64, error) {
```
Comment (dcrystalj) [inline plugins/boltz/swap.go:29]: FundsToReceive => onchain or offchain ?

File: plugins/boltz/db.go (hunk 1 above)
Comment (fiksn) [reply plugins/boltz/db.go:26]: You're right, changed that to create directory (which does not necessary exist)

File: plugins/boltz/swap.go (hunk 2 above)
Comment (fiksn) [reply plugins/boltz/swap.go:30]: Yeah CalcFundsToReceive will be refactored.

File: plugins/boltz/swap.go [hunk 5]
```diff
@@ -4,10 +4,94 @@ import (
...
 
 	"github.com/BoltzExchange/boltz-lnd/boltz"
 )
 
+func (b *Plugin) GetStatus(boltzID string, reverse bool) {
+
+	b.BoltzAPI.SwapStatus(boltzID)
+
+	/*
+			transaction.confirmed: that transaction was included in a block
```
Comment (dcrystalj) [inline plugins/boltz/swap.go]: weird indent

File: plugins/boltz/fsm.go [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+	in.SwapData.State = state
+	err := in.BoltzPlugin.db.Insert(in.SwapData, in.SwapData.JobID)
+	if err != nil {
+		in.MsgCallback(entities.PluginMessage{
+			JobID:      int32(in.GetJobID()),
+			Message:    fmt.Sprintf("Could not change state to %v %v", state, err),
+			IsError:    true,
+			IsFinished: true,
+		})
+	}
```
Comment (dcrystalj) [inline plugins/boltz/fsm.go:129]: I would send state successful transition to our backend as well.

File: plugins/boltz/job.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+const (
+	ErrCouldNotParseJobData  = Error("could not parse job data")
+	ErrUnsupportedTargetType = Error("unsupported target type")
+)
+
+type TargetType string
+
+const (
+	OutboundLiqudityNodeTarget   TargetType = "OutboundLiqudityNodeTarget"
```
Comment (dcrystalj) [inline plugins/boltz/job.go:19]: OutboundLiqudityNodePct ?

File: plugins/boltz/main.go [hunk 8]
```diff
@@ -109,17 +123,24 @@ func NewPlugin(lnAPI agent_entities.NewAPICall, filter filter.FilteringInterface
 		if err != nil {
 			return nil, err
 		}
+		err = db.Insert(SecretDbKey, &Entropy{Data: entropy})
```
Comment (dcrystalj) [inline plugins/boltz/main.go:126]: looks like mnemonic setup could be extracted to new function. 
I would love to have NewPlugin() as clear as possible

File: plugins/boltz/main.go [hunk 9]
```diff
@@ -128,24 +149,52 @@ func NewPlugin(lnAPI agent_entities.NewAPICall, filter filter.FilteringInterface
...
 	}
 
+	resp.MaxFeePercentage = cmdCtx.Float64("maxfeepercentage")
+
 	return resp, nil
 }
 
-func (b *Plugin) Execute(jobID int32, data []byte, msgCallback entities.MessageCallback) error {
-	jd := &types.JobData{}
+func (b *Plugin) jobDataToSwapData(jobData *JobData) *SwapData {
```
Comment (dcrystalj) [inline plugins/boltz/main.go:167]: pass context,
move under Execute()

File: plugins/boltz/normal_fsm.go [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+)
+
+const (
+	BtcPair = "BTC/BTC"
+	Btc     = "BTC"
+)
+
+// Normal (submarine) swap finite state machine
+
+func FsmInitialNormal(in FsmIn) FsmOut {
```
Comment (dcrystalj) [inline plugins/boltz/normal_fsm.go:21]: Lets try to avoid `normal` terminology and stickt to swap-in / out or submarin/reverse-submarine

**Review by dcrystalj — APPROVED** [review]

File: plugins/boltz/redeemer.go [hunk 11]
```diff
@@ -88,24 +103,61 @@ func (r *Redeemer) redeem() bool {
...
+		return true
+	}
+
 	outputs := make([]boltz.OutputDetails, 0)
 	used := make([]JobID, 0)
 	for _, entry := range r.Entries {
-		output := r.getRefundOutput(&entry)
+		var output *boltz.OutputDetails
+		sd := entry.GetSwapData()
+		if sd.State == RedeemingLockedFunds {
```
Comment (dcrystalj) [inline plugins/boltz/redeemer.go:126]: use switch

File: plugins/boltz/swapmachine.go [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+	MsgCallback entities.MessageCallback
+}
+
+// To satisfy interface
+func (i FsmIn) GetSwapData() *SwapData {
+	return i.SwapData
+}
+
+// changeState actually registers the state change
+func (b *Plugin) changeState(in FsmIn, state State) error {
```
Comment (dcrystalj) [inline plugins/boltz/swapmachine.go:28]: b -> p

File: plugins/boltz/swapmachine.go [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+			JobID:      int32(in.GetJobID()),
+			Message:    msg,
+			IsError:    false,
+			IsFinished: false,
+		})
+	}
+}
+
+func (s *SwapMachine) getSleepTime(in FsmIn) time.Duration {
+	if s.BoltzPlugin.ChainParams.Name == "mainnet" {
```
Comment (dcrystalj) [inline plugins/boltz/swapmachine.go:59]: duplicated code from boltz/main.go

File: plugins/boltz/swapmachine.go [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+		return 1 * time.Minute
+	}
+	return 5 * time.Second
+}
+
+func (s *SwapMachine) FsmNone(in FsmIn) FsmOut {
+	panic("Invalid state reached")
```
Comment (dcrystalj) [inline plugins/boltz/swapmachine.go:66]: we should also send to backend this message

File: plugins/boltz/swapmachine.go [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+			IsFinished: true,
+		})
+	}
+	return FsmOut{}
+}
+
+func (s *SwapMachine) RedeemedCallback(data FsmIn, success bool) {
+	sd := data.GetSwapData()
+	if sd.State == RedeemingLockedFunds {
+		// even when this succeeds state is still screwed up
```
Comment (dcrystalj) [inline plugins/boltz/swapmachine.go:96]: why? please add more explanation

File: plugins/boltz/swapmachine.go [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+		if success {
+			s.Eval(data, SwapFailed)
+		}
+	} else if sd.State == ClaimReverseFunds {
+		if success {
+			s.Eval(data, SwapSuccess)
+		} else {
+			s.Eval(data, SwapFailed)
+		}
+	}
```
Comment (dcrystalj) [inline plugins/boltz/swapmachine.go:106]: else error/panic ?

File: plugins/boltz/swapmachine.go [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+func NewSwapMachine(plugin *Plugin) *SwapMachine {
+	s := &SwapMachine{
+		Machine: &Fsm[FsmIn, FsmOut, State]{States: make(map[State]func(data FsmIn) FsmOut)},
+		// TODO: instead of BoltzPlugin this should be a bit more granular
+		BoltzPlugin: plugin,
+	}
+
+	s.Machine.States[SwapFailed] = FsmWrap(s.FsmSwapFailed, plugin)
```
Comment (dcrystalj) [inline plugins/boltz/swapmachine.go:191]: do mapping in one op, at line 186

Comment (fiksn) [reply plugins/boltz/swapmachine.go:191]: Thought about that, but can't because function depends on `s` instance.

File: plugins/boltz/e2e_test.go [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+		api, err := api.NewAPI(api.LndGrpc, func() (*common_entities.Data, error) {
+			return data, nil
+		})
+
+		return api, err
+	}
+}
+
+func msgCallback(msg agent_entities.PluginMessage) error {
+	fmt.Printf("%+v\n", msg)
```
[the remark this answers is not in the document]
Comment (fiksn) [reply plugins/boltz/e2e_test.go]: Good idea, but this is not structured enough, perhaps just if we find a message containing substring success or sth.

File: plugins/boltz/swapmachine.go [hunk 19]
```diff
@@ -56,13 +56,11 @@ func log(in FsmIn, msg string) {
...
 func (s *SwapMachine) getSleepTime(in FsmIn) time.Duration {
-	if s.BoltzPlugin.ChainParams.Name == "mainnet" {
-		return 1 * time.Minute
-	}
-	return 5 * time.Second
+	return s.BoltzPlugin.getSleepTime()
 }
 
 func (s *SwapMachine) FsmNone(in FsmIn) FsmOut {
+	log(in, "Invalid state reached")
```
Comment (dcrystalj) [inline plugins/boltz/swapmachine.go:63]: shouldnt we set `IsFinished: true` ?

**Review by dcrystalj — APPROVED** [review]

File: plugins/boltz/swapmachine.go (hunk 19 above)
Comment (fiksn) [reply plugins/boltz/swapmachine.go:63]: I am not sure about the `log` already. `None` is not a state you could encounter during regular flow. So if you hit it, it doesn't mean swap has failed, it means the version of agent is somehow totally broken. So I guess you don't even want to transition to `None` state but just crash. The logic behind is that if you change agent binary to some newer build you might catch up from the last regular state and still bring the swap to a final state.
